spec/views/produce_items/index.html.erb_spec.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'produce_items/index', type: :view do
```

Comment:
I struggle a bit to find utility in view tests. I think they are fine for training but they don't add a lot of value. I'd say that adding e2e tests in the future would be a better bet. WDYT?

Reply:
What kind of tests do you find usefull? Do you think we shouldn't have any view tests at all? Regarding models and controllers tests, would you remove/add something? Ho would you go abtou writing e2e tests?

Reply:
Tests are a topic in which opinions vary a lot. I'll start on what (almost) everyone agrees and then highlight the places in which I have a different perspective.

So we all usually use what is called the test pyramid:

![](https://martinfowler.com/articles/practical-test-pyramid/testPyramid.png)

The quantity of tests you should have is equal (in proportion) to the width of the step. Thus we should have mostly unit tests and few e2e tests.

Now what is a unit tests is a controversial topic, the classical definition states that these tests are ISOLATED, meaning there is no interaction with anything but the piece of code you are testing. This is not applicable to Rails, since we do not mock the database to make our model tests (although we still call them "unit tests"). So from my perspective, IN A RAILS application, I use the following definitions:

- Unit tests: tests that test a single class. Database interactions are fine as long we're not testing two different objects
- Integration tests: tests that test a combination of classes or contain side effects. Testing controllers is a good example, but there are also others, like testing a [service object](https://www.toptal.com/ruby-on-rails/rails-service-objects-tutorial) that joins several other service objects
- e2e tests: tests that actually boot a browser, click on stuff and test behavior (`capybara` does those, `cypress` does them as well)

Now there is the question of what should be tested:

- Everything inside models must be tested
- Every service object must be tested
- Controller tests are painful to maintain (and also made useless by e2e tests). We should only make these to test API endpoints
- View tests are useless if we have e2e tests. Testing what is rendered has little value IMO
- e2e tests are very important, but should be created with reservation, since they are very very slow. However we should test all our important flows (meaning user stories/scenarios).

Hope that sheds some light into this topic :x

OBS.: Read [this article](https://martinfowler.com/articles/practical-test-pyramid.html) to get a more traditional perspective